Introducing Repair system

pave1-semenov opened the pull request:
# Introducing Repair system

The system is described at p.140 of Corebook and is based on the following rules:

 - In order to fix an item you need **exactly one** unit of each crafting material listed in the original diagram
 - You can only fix a weapon or armor
 - To fix the item you must make a **Crafting roll** at a DC equal to the crafting DC stated in diagram minus 5. For every glyph/rune/enchant the DC is increased by 2
 - If you fail the check - the item remains broken and all of the used materials are gone. However you can attempt a recovery check with the same DC to save the materials (**this is not implemented**)
 - If you succeed - the item regains its full durability

## Assumptions
As it's not directly stated in the corebook I assume that:

 - Repairing the item always requires the same amount of materials no matter how broken it is
 - It doesn't matter if the item has a lot of damaged locations - you always use 1 unit of each material. E.g. fixing the full-cover armor with 6 damaged/broken parts is the same as fixing the sword with a single reliability stat
 
https://github.com/user-attachments/assets/ec6f1e43-b3ae-4284-b9b5-8ea9af6d3280

## Features
The system is mostly based on the existing Crafting system, but inroduces some new features.
### Associated diagrams
In order to repair a weapon or armor in needs to have an associated crafting diagram. This is required to pull the Crafting DC and the list of crafting materials from the diagram. 
You can find the new Crafting diagram section in the weapon and armor sheets. It behaves pretty much the same as the associated items in the crafting diagrams themselves.

![image](https://github.com/user-attachments/assets/ac0f976f-a429-4097-808b-fbc05ead5c85)
![image](https://github.com/user-attachments/assets/e8679499-9d08-4eeb-a59a-f4a959c69eea)

Upon adding a diagram the system would check if the diagram type matches the item type (e.g. only weapon diagrams could be associated with weapons)

### Repair
The new repair icon could be found near the Delete item icon in the Gear & Inventory tab -> Equipped Weapons/Armor -> Info.

![image](https://github.com/user-attachments/assets/997dca5c-ac8b-4259-b1db-dee4290389b3)

The icon is not always visible as there are some conditions which needs to be met:

 - Item must have an associated diagram
 - Item must be damaged. For the weapons and shields it means that current reliability is lower than max relibility, for the armor - any of the locations doesn't have a max SP
#### Repair modes
Clicking the repair icon would bring a repair dialog which displays the summary of the work:
 - Item description
 - Repair DC
 - Damaged locations with current and max reliability
 - The list of required materials and the quantity character currently carry
 
![image](https://github.com/user-attachments/assets/0e7a29dd-662e-4995-adb6-ed8cd9cc80e7)

You can than use one of the buttons corresponding to the repair modes:
##### Repair
Just getting the job done. The Craftring roll would be made for the owner of the item and the materials would be subsctracted from the charcter's stocks. You can't use this mode if you don't have enough materials to repair the item.

https://github.com/user-attachments/assets/e7e50ed4-d1a9-4386-a05c-681bd971efbc

##### Simulate repair
Based on the crafting with no 'real craft' mode selected. Would just make a Crafting roll without substracting any materials and restoring the durability.
Added this as a backward compatibility to the original crafting. I'm not really sure it's required for the repair system though

https://github.com/user-attachments/assets/394cffaf-d8f4-4d04-8479-8d541eecbcdb

##### Request repair
It's pretty common that not every player could repair their own gear and it's better to leave it to the professionals. Especially if you have an artisan in a party.
Requesting repair would send an 'order' to the chat from where anyone could take it and do the job using their own materials and Crafting roll. Total repair price would be calculated based on the crafting material cost.

https://github.com/user-attachments/assets/d9d55e82-f745-4bfb-beab-637e7d115c86

##### GM repair
This mode is only available for GMs when fulfilling the repair request. Repair automatically succeeds and no crafting components are consumed. You can use this mode even if you don't have enough components. Total repair price is calculated based on the individual components cost.
However, there is a drawback with a price calculation - if the actor who performs the repair doesn't have some of the components in their inventory - the price for those components won't be available. GM can specify the cost of the missing components manually and it will be added to the total price.

https://github.com/user-attachments/assets/ce249959-4691-4d51-b803-5f34e08abc6a

Stexinator commented:
Can you add that when "requesting repair" you dont have to roll for npcs?

Stexinator commented:
Hey, you got that correct. Gold plating would be to make it an GM only button and calculate the cost.

Thanks for the enhanced description

Stexinator commented on module/actor/sheets/WitcherCharacterSheet.js:
```diff
@@ -305,4 +307,11 @@ export default class WitcherCharacterSheet extends WitcherActorSheet {
+        let itemId = event.currentTarget.closest(".item").dataset.itemId
+        let item = this.actor.items.get(itemId)
+
+        await RepairSystem.process(this.actor, item)
```
I think repairing an item should live on the item.

so calls should look like ``item.repair()`` etc

Stexinator commented on module/data/item/templates/associatedItemData.js:
```diff
@@ -0,0 +1,15 @@
+    const field = {}
+    field[fieldName] = new fields.StringField({initial: ''})
+
+    return field
```
I dont think a generic method in the data template is good. I would like for you to make it explicit that only associatedDiagrams are added

Stexinator commented on module/item/sheets/WitcherArmorSheet.js:
```diff
@@ -1,5 +1,6 @@
 import WitcherConfigurationSheet from './configurations/WitcherConfigurationSheet.js';
 import WitcherItemSheet from './WitcherItemSheet.js';
+import { associatedDiagramMixin } from '../../actor/sheets/mixins/associatedDiagramMixin.js';
```
item sheets should use their own mixins. Here you access the mixins of the actor

Stexinator commented on module/item/sheets/WitcherWeaponSheet.js:
```diff
@@ -1,5 +1,6 @@
 import WitcherDamagePropertiesConfigurationSheet from './configurations/WitcherDamagePropertiesConfigurationSheet.js';
 import WitcherItemSheet from './WitcherItemSheet.js';
+import { associatedDiagramMixin } from '../../actor/sheets/mixins/associatedDiagramMixin.js';
```
same

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -0,0 +1,301 @@
+    { label: ['WITCHER.Repair.damagedLocations.shield'], reliability: 'reliability', maxReliability: 'reliabilityMax' },
+]
+
+const repairModifier = 5
```
the modifier should be -5 making it implict at - in the formula might lead to confusion in the long term

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -0,0 +1,301 @@
+
+        let rollFormula = displayRollDetails
+            ? `1d10+${stat}[${statName}]+${skill}[${data.skillName}]`
+            : `1d10 + ${stat}`;
```
Here you forgot the skill

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -0,0 +1,301 @@
+    }
+
+    get skillName() {
+        return game.i18n.localize(this.executor.system.skills.cra.crafting.label).replace(" (2)", "")
```
You can query this from the config

Stexinator commented on module/scripts/item/updateItem.js:
```diff
@@ -0,0 +1,14 @@
+
+export function updateItemByUuid(itemId, data) {
+    const item = fromUuidSync(itemId)
+    item?.update(data)
```
I didnt add generic update calls to prevent abuse and user errors. Allowing anyone to just update an item with any data is prone to produce problems in the long term

Stexinator commented on templates/dialog/repair-dialog.hbs:
```diff
@@ -0,0 +1,66 @@
+<div class="flexcol">
+    <div class="scrollable">
+        <header class="item-header">
+            <h1 class="itemname">Repairing {{data.item.name}}</h1>
```
Here is a localize missing

pave1-semenov commented:
Hello there! 
I made the fixes according to the comments - please take a look.

I also implemented the 'GM repair' mode as we discussed and updated the description with a screencast.

There is, however, a drawback with the price calculation. Since crafting diagrams don’t have actual associated components, it’s difficult to determine the cost if the actor doesn’t have them in their inventory. To address this, I’ve added input fields for each missing component, allowing costs to be entered manually. These manual entries will be included in the total price.
I plan to address this issue in the next iteration by making components linkable to diagrams. With this feature, it would become possible to implement armor and weapon disassembly.

Stexinator commented:
https://github.com/witchertrpg-foundryvtt/TheWitcherTRPG/issues/184 here is a macro for price calculation

Stexinator commented on lang/en.json:
```diff
@@ -1241,13 +1241,17 @@
                 "repair": "Repair",
                 "simulate": "Simulate repair",
-                "request": "Request repapir"
+                "request": "Request repapir",
```
typo

Stexinator commented on lang/en.json:
```diff
@@ -1241,13 +1241,17 @@
+                "request": "Request repapir",
+                "gmRepair": "GM repair"
             },
             "dialog": {
```
I think the whole bloch should be moved to "repair" or something

pave1-semenov replied:
You mean adding a seprate key for the lowercase "repair" to reuse it in buttons?

Stexinator replied:
You put it under "dialog" but as far as I can see all the things are used for repair

pave1-semenov replied:
These are already under repair block - path for these strings is like `WITCHER.Repair.buttons.gmRepair`

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -140,7 +150,7 @@ class Repair {
     }
 
-    async repairItem(data, simulate) {
+    async repairItem(data, simulate, gmRepair) {
```
Due to readability I would prefer that ``simulate`` and ``gmRepair`` would be wrapped in an ``options``-object

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -150,6 +160,32 @@ class Repair {
+        html.find('.component-cost').on('change', this._onComponentCostChange.bind(this, data))
+    }
+
+    _onComponentCostChange(data, event) {
```
Diagrams have a list of components and these components also have a price attached. See the link to the macro I posted

pave1-semenov replied:
Yeah, that's what I use for 'ownedComponents' which are in the actors inventory - see `RepairData.repairPrice`.  
These inputs are only for the missing components - I wanted to avoid the compendium lookups in this case since the compendium is not a hard dependency of the system. 
Also the system allows to create your own diagrams with any components - not necessarily from compendiums

Stexinator replied:
Could you add a check if the compendium is present and if not query the user for prices?

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -274,25 +339,29 @@ class RepairData {
-    get repairDCForumulae() {
-        let formulae = `${this.diagram.system.craftingDC}[${game.i18n.localize('WITCHER.Repair.params.craftingDC')}] - ${repairModifier}[${game.i18n.localize('WITCHER.Repair.params.repairMod')}]`
+    get repairDCForumula() {
+        let formula = `${this.diagram.system.craftingDC}[${game.i18n.localize('WITCHER.Repair.params.craftingDC')}] - ${Math.abs(repairModifier)}[${game.i18n.localize('WITCHER.Repair.params.repairMod')}]`
```
instead of using Math.abs you can just add everything up

pave1-semenov replied:
But this is a string formula - putting '+' would interpolate into the '+' literal

Stexinator replied:
Ah I see. Then this is fine for me

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -274,25 +339,29 @@ class RepairData {
-        let formulae = `${this.diagram.system.craftingDC}[${game.i18n.localize('WITCHER.Repair.params.craftingDC')}] - ${repairModifier}[${game.i18n.localize('WITCHER.Repair.params.repairMod')}]`
+    get repairDCForumula() {
+        let formula = `${this.diagram.system.craftingDC}[${game.i18n.localize('WITCHER.Repair.params.craftingDC')}] - ${Math.abs(repairModifier)}[${game.i18n.localize('WITCHER.Repair.params.repairMod')}]`
         const echantsCount = this.enchantsCount
```
typo

pave1-semenov replied:
Which one? Thought I corrected the typo here - was an unnecessary plural form

Stexinator replied:
Forumula, echantsCount

Stexinator commented on module/item/systems/repair.js:
```diff
@@ -274,25 +339,29 @@ class RepairData {
 
     get skillName() {
-        return game.i18n.localize(this.executor.system.skills.cra.crafting.label).replace(" (2)", "")
+        return game.i18n.localize(WITCHER.skillMap.crafting.label)
```
The config should be loaded from CONFIG.WITCHER... so that modules or world scripts can modify stuff

pave1-semenov replied:
This is ineed the `CONFIG.WITCHER` or am I loading it incorrectly?
`import { WITCHER } from "../../setup/config.js";`

Stexinator replied:
You are loading the system defined WITCHER config. In the setup this is added to the global CONFIG object where modules or world scripts can change things.

Stexinator replied:
![grafik](https://github.com/user-attachments/assets/4a4d154f-cb07-4ce0-a364-ef6a78453c11)

pave1-semenov replied:
This was fixed in one of the latest commits.
Forgot to resolve it, sorry

pave1-semenov commented:
Hi @Stexinator! 
I added the `compendiumLookup` function based on your macro. Now the manual price edit is only available for the 'unknown components` (could not be found neither in inventory, nor in compendium).
Please take a look.

Stexinator commented:
@pave1-semenov are you still working on this?

pave1-semenov commented:
> @pave1-semenov are you still working on this?

No, all the threads seems to be resolved or am I missing something?

Stexinator commented:
> > @pave1-semenov are you still working on this?
> 
> No, all the threads seems to be resolved or am I missing something?

There is still the issue with the wrong source for the skill lookup. I could fix that on dev if you dont know how to resolve it.

Can you squash your commits so that only meanigful commit messages remain?

Stexinator commented:
@pave1-semenov I dont know if you followed the v13 prototype. But they decided that all v1 apps will be unsupported with v16. You dont need to move everything to AppV2 in this comment but I would like if if you could do it in a follow-up PR

Stexinator commented:
When you are ready to merge you can request a re-review so that I know that you are done

pave1-semenov commented:
> @pave1-semenov I dont know if you followed the v13 prototype. But they decided that all v1 apps will be unsupported with v16. You dont need to move everything to AppV2 in this PR but I would like if if you could do it in a follow-up PR

Noted 👍

Stexinator commented:
You got a typo in a commit message. If you dont care I can merge it

pave1-semenov commented:
> You got a typo in a commit message. If you dont care I can merge it

Fixed now

Stexinator commented:
Can you do a rebase on dev? Only complete versions are merged via merge commit. Thanks

pave1-semenov commented:
> Can you do a rebase on dev? Only complete versions are merged via merge commit. Thanks

Done 👍
